src/vsc_r_api.ts

```diff
@@ -0,0 +1,228 @@
+}
+
+export async function showDialog(message: string) {
+
+  window.showInformationMessage(message);
+
+}
+
+export async function navigateToFile(file: string, line: number, column: number) {
+
+  const targetDocument = await workspace.openTextDocument(Uri.file(file));
+  const editor = await window.showTextDocument(targetDocument);
+  if (line > 0 && column > 0) {
+    const targetPosition = parsePosition([line, column]);
+    editor.selection = new Selection(targetPosition, targetPosition);
+  }
+}
+
+export async function setSelections(ranges: number[][], id: string) {
+  // In VSCode it's not possible to get a list of the open text editors. it is
+  // not window.visibleTextEditors - this is only editors (tabs) with text
+  // showing. So we have to open the target document and and 'show' it, to get
+  // access to the editor object and manipulate its' selections. This is
+  // different from RStudio which can manipulate the selections and cursor
+  // positions in documents on open tabs, without showing those documents.
+  const target = findTargetUri(id);
+  const targetDocument = await workspace.openTextDocument(target);
+  const editor = await window.showTextDocument(targetDocument);
+
+  const selectionObjects = ranges.map(x => {
+    const newRange = parseRange(x);
+    const newSelection = new Selection(newRange.start, newRange.end);
+    return (newSelection);
+  });
+
+  editor.selections = selectionObjects;
+}
+
+export async function documentSave(id: string) {
+  const target = findTargetUri(id);
+  const targetDocument = await workspace.openTextDocument(target);
+  await targetDocument.save();
+}
+
+// interface
+
+// represents addins in a QuickPick menu
+interface AddinItem extends QuickPickItem {
+  binding: string;
+  package: string;
+}
+
+// This a memoised style function that only generates the list of addins for the 
+// quick pick menu once, returning the generated array on every subsequent call. 
+export const getAddinPickerItems = (() => {
+
+  let addinQuickPicks: AddinItem[] = undefined;
+
+  return async () => {
+    if (typeof addinQuickPicks === 'undefined') {
+      const addins: any[] = await fs.readJSON(path.join(sessionDir, 'addins.json'));
+      const addinItems = addins.map((x) => {
+        return {
+          alwaysShow: true,
+          description: x.description,
+          label: x.name,
+          detail: x.package,
+          picked: false,
+          binding: x.binding,
+          package: x.package,
+        };
+      });
+      addinQuickPicks = addinItems;
+    }
+    return addinQuickPicks;
+  };
+})();
+
+export async function launchAddinPicker() {
+
+  if (!config().get<boolean>('sessionWatcher')) {
+    throw ('{rstudioapi} emulation requires session watcher to be enabled in extension config.');
+  }
+  if (typeof rTerm === 'undefined') {
+    throw ('No active R terminal session, attach one to use RStudio addins.');
+  }
+
+  const activeRTerm = await chooseTerminal();
+
```

Comment:
I'm thinking that we might implement a little function of wait-for-attach and await it here.

The function could return a `Promise` where `sessionDir` is constantly checked or timeout.

Reply:
That's pretty much exactly what I had in mind. 👍

Reply:
Currently, we don't reset variables like `pid` and `sessionDir` when session is terminated because if a session is killed we don't know that anyway. Resetting `sessionDir` before await attach seems necessary.